### PR TITLE
Handle HTTP 429 responses with a request backoff approach

### PR DESCRIPTION
**Pull request by muesli**
Since it's difficult to wrap all possible go-mastodon API calls in a backoff
algorithm outside of the package itself, I decided to implement a simple
version of it in go-mastodon's doAPI itself.

This works nicely, but could be improved in two ways still:

- Abort sleeping when context gets cancelled
- Make backoff optional / configurable

Personally, I still think this is a good start and probably fits most of
go-mastodon's use-cases. It certainly beats string-grepping for status code
"429" in clients.

### CODE REVIEW

**Comment by muesli** [comment]
@mattn Gentle ping :smile: Any feedback you could provide on this PR?

File: mastodon.go [hunk 1]
```diff
@@ -98,11 +99,28 @@ func (c *Client) doAPI(ctx context.Context, method string, uri string, params in
...
+
+		// handle status code 429, which indicates the server is throttling
+		// our requests. Do an exponential backoff and retry the request.
+		if resp.StatusCode == 429 {
+			if backoff > time.Hour {
+				break
+			}
+
+			backoff *= 2
+			time.Sleep(backoff)
```
Comment (mattn) [inline mastodon.go]: Should check ctx.Done() ?
```go
select {
case time.After(backoff)
case <-ctx.Done():
	return ctx.Err()
}
```

Comment (muesli) [reply mastodon.go]: Sorry this took so long, I must have accidentally deleted the GitHub notification.

Just pushed pretty much what you proposed, which handles the context cancellation.

**Comment by muesli** [comment]
@mattn Ready for another review!

**Comment by mattn** [comment]
Thank you

**Pull request merged**